### ISSUE TITLE
Add a "Porchfile" to kpt packages that stores porch-specific metadata

### ISSUE THREAD

**Issue by kispaljr** [nephio-project/nephio#783]
### TL;DR

 Add a YAML file named `Porchfile` to kpt packages managed by porch to persistently store porch-specific metadata. 

### Overview

Similarly to how kpt stores package metadata in Kptfile, porch should store its own package metadata in Porchfile. Porchfile should contain a valid KRM object, with the local-config annotation. Here is a possible example Porchfile content:
```YAML
apiVersion: porch.kpt.dev/v1
kind: Porchfile
metadata:
  annotations:
    config.kubernetes.io/local-config: "true"
  name: my-kpt-package
spec:
  readinessGates:
    - req.nephio.org/v1alpha1.Interface.n3
    - workload.nephio.org/v1alpha1.NFDeployment.upf-cluster01
status:
  conditions:
  - message: update condition for initial resource
    reason: workload.nephio.org/v1alpha1.NFDeployment.upf-cluster01
    status: "False"
    type: req.nephio.org/v1alpha1.Interface.n3
  - message: update for condition
    status: "False"
    type: workload.nephio.org/v1alpha1.NFDeployment.upf-cluster01
  lastRender:
    error: ""
    result:
      exitCode: 0  
```

### What kind of metadata should be stored in Porchfile?

Let's assume that we have a kpt package in a git repository, and we have multiple porch instances running in separate Kubernetes clusters that has the same git repository registered. 

Some of the metadata belonging to the package should be always the same no matter via which porch instance we observe the package. Or, in other words, if this kind of metadata is changed via one porch instance the change should be visible in all other porch instances. Some examples for this kind of metadata: 
- description
- upstream package
- mutation pipeline
- readiness gates
- conditions
- results of the latest render operation performed by porch

These kinds of metadata should be stored either in `Kptfile` - if it is owned by kpt, or in `Porchfile` - if the data is owned by porch.

There are other kinds of package metadata that shouldn't be synchronized across porch instances, e.g.:
- owner references
- labels
- annotations

These types of package metadata should be persisted in the so-called "metadata store" that is already implemented in porch [here](https://github.com/nephio-project/porch/blob/main/pkg/meta/store.go).

Historically kpt and porch were part of the same project, so there were no need for a separate `Kptfile` and `Porchfile`. This is the reason for example that `readinessGates` and `conditions` that are only used by porch became part of Kptfile. Those two should be migrated from Kptfile to Porchfile.

#### Kptfile

A natural alternative storage for the Porchfile metadata would be the Kptfile itself. Unfortunately that would make porch-managed kpt packages incompatible with the `kpt` CLI tool. This is a problem for porch developers and users, since we still heavily utilize the kpt command for development and troubleshooting of kpt packages, e.g. to run `kpt fn render` on a local copy of a kpt package, or deploying a kpt package by `kpt live apply`.

#### Metadata store

We already established above that the porch-internal metadata store was designed to store other kinds of metadata than is proposed to be stored in a Porchfile.

**Comment by kispaljr** [comment]
cc: @henderiw  @tliron @johnbelamaric @Catalin-Stratulat-Ericsson @efiacor @liamfallon

**Comment by henderiw** [comment]
Ok lets firsts try to highlight the problem we want to solve as this is not clear from this issue.

**Comment by kispaljr** [comment]
I will try to rephrase it: 
We want to store extra information that is related to a particular kpt package, but that is not strictly its contents. All such information/metadata is now stored in the Kptfile (ref to the upstream package, readinessGates, conditions, etc.). Now we encountered a situation, when we want to store a new kind of metadata needed for porch to do its job properly, but we cannot just add a new field to the Kptfile, since kpt is a separate project now and they own the schema of the Kptfile.

So according to this proposal porch will to create its own file that is similar to Kptfile, but separate from it, and whose schema is fully owned by the porch project. This will maintain compatibility between kpt packages managed by porch and the kpt CLI command, and still allow us a place to introduce new kind of data that we want to store ina package.

**Comment by henderiw** [comment]
Why does it have to be a file and not just metadata to the PR CRD?

**Comment by kispaljr** [comment]
That is what I tried to justify in the original wording of the PR, by mentioning the use case when e.g. a vendor's blueprint repo is registered in multiple customers' porch instances.
But another good example would be if you unregister a repository from porch (by deleting the Repository object) and re-register it. All labels and annotation added to the PackageRevision CRs will be lost after re-registering the repo. And this is fine, but there is types of package metadata that you don't want to lose when you unregister the repo from porch, for example: the reference to the upstream package, or the readinessGates. This kind of metadata should stored along with the package in git. And the easiest way to add this information to git is to create a file. I would say that this is reasoning behind having a Kptfile in the first place.

**Comment by henderiw** [comment]
There is some conflicting info here:
1. you are talking about a blueprint, you don't want specific information stored in a common blueprint repo.

and some info specific to porch implementation.
1. the fact you loose metadata  if you deregister a repo it is a bug in porch. The PR is a way to convey information to the user about the result of things that happened. I don't believe it is good practice to store metadata in git. The fact we use the kptfile to convey information like readiness is also not the right way.

**Comment by kispaljr** [comment]
Yes. I agree, but the backend storage behind PackageRevision objects is git, isn't it? Similarly how normal CRs are stored in etcd. 

As far as I can see, all persistent data presented via PackageRevision objects has to be somehow stored in git. Either encoded in the name of a git branch (how we store workspace names for draft/proposed packages), or in git annotations (like how store workspace names for published packages), or in files in git (that's what we do with readinessGates in Kptfile), or maybe using some other git concept, but definitely in git. 
A,m I missing something?

**Comment by kispaljr** [comment]
For the record: I completely rewrote the original proposal text in the first comment, based on the discussion with @henderiw . I hope the new version is a more conscious and digestible description of what we want to add and why.

**Comment by tliron** [comment]
I support this idea and indeed implemented it in the [TKO experiment](https://github.com/nephio-experimental/tko/blob/main/PACKAGES.md#deploymentnephioorg), where it's used to keeps track of bookkeeping metadata (URL of the original template used) as well as allowing for some optional user-controlled knobs for changing default behavior, such as disabling auto-approval ([example](https://github.com/nephio-experimental/tko/blob/main/examples/free5gc-smf/deployment.yaml)).

I don't think "Porchfile" is the best name, though, as hopefully it can be agnostic to any specific implementation.

**Comment by kispaljr** [comment]
By all means, let's find a better name for this file :)

**Comment by henderiw** [comment]
this is the proper way to solve this, rather than keep hacking. My 2 cents.
https://github.com/nephio-project/nephio/issues/598

**Comment by kispaljr** [comment]
I see this as a different, independent problem.
I assume even if PackageRevisions are CRs, we want to keep registering Registrys and auto-discovering the packages in them (there will be a PackageRevision controller that creates/updates/deletes PackageRevisions based on what is in git). If this is true, then we still have to decide where the "persistent metadata" attached to a package is stored between Registry unregistration and re-registration.  In other words we have to decide where the PackageRevision controller will read the persistent metadata values from, during auto-discovery.
Porchfile tries to solve the problem of persisting metadata whose lifecycle is bound to the lifecycle of a package in git (that I called "persistent metadata", and not the problem of storing metadata whose lifecycle is bound to the lifecycle of PackageRevision objects (that I called "ephemeral metadata").

**Comment by henderiw** [comment]
Ok I reread your new text, but the way to solve this is a bit different in my view and I am of the opinion porch should just care about access to git, not about any specialization, etc. As such a workspace name is a local matter, so it does not need to be persisted and it does not need to be coordinated across clusters.
Wrt condiitons, etc that work is triggered from a PV and the corresponding technologies should report their status, so the reproducibility should come from that chain not porch.
